Use AnimatedSwitcher's _childNumber as Key in layoutBuilder's Stack

jimgerth opened the pull request:
Previously `KeyedSubtree.wrap` gave children of the `AnimatedSwitcher` their own key as a key, if available, and their `_childNumber` as a key, if not. This could lead to multiple children having the same key in the `defaultLayoutBuilder`'s `Stack` and thus a duplicate key exception being thrown, though (see #121336 for a detailed description of how this bug occurs). This is fixed by always using the uniquely identifiable `_childNumber` as a key for the children of the `AnimatedSwitcher`.

This does not impair the functionality of having children of the same type be differentiated and transitioned between by giving them different keys, as the children's original key is still used to determine if a transition should take place or not.

This fixes #121336.

**Note:** This reverts the changes introduced by #107476, as they do not completely fix the issue of duplicate keys in the AnimatedSwitcher switcher as demonstrated by #121336, and introduce behavior that deviates from the AnimatedSwitcher's documentation.
Specifically, children that are transitioning out are immediately removed from the widget tree, if another child with the same key is added, while the AnimatedSwitcher is clearly supposed to support multiple children with the same key transitioning in and out at the same time according to its documentation: 
> The same key can be used for a new child as was used for an already-outgoing child; the two will not be considered related. (For example, if a progress indicator with key A is first shown, then an image with key B, then another progress indicator with key A again, all in rapid succession, then the old progress indicator and the image will be fading out while a new progress indicator is fading in.)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

jimgerth committed:
Use AnimatedSwitcher's _childNumber as Key in layoutBuilder's Stack

Previously KeyedSubtree.wrap gave children of the AnimatedSwitcher their own key
as a key, if available, and their _childNumber as a key, if not. This could lead
to multiple children having the same key in the defaultLayoutBuilder's Stack and
thus a duplicate key exception being thrown, though.
This is fixed by always using the uniquely identifiable _childNumber as a key
for the children of the AnimatedSwitcher.

This does not impair the functionality of having children of the same type be
differentiated and transitioned between by giving them different keys, as the
children's original key is still used to determine if a transition should take
place or not.

goderbauer commented:
@JimGerth Can you please sign the CLA (see previous message) and add a test for your change? Also, it looks like some other checks are failing as well. Please take a look at those as well. Thanks.

JimGerth commented:
@goderbauer I have signed the CLA now and will add a test for my change once I have time.

Also it seems that the test failing currently is the one introduced by #107476, fixing a related issue of duplicate keys in the `AnimatedSwitcher` in what seems to me a non-optimal way.

Children, that are transitioning out are simply completely removed from the widget tree, if another child with the same key is added. As the test's name introduced by that PR implies:
> AnimatedSwitcher does not duplicate animations if the same child is entered twice.

_(And this does not even completely eliminate that problem of duplicate keys, as shown by #121336, which is why this PR exists...)_

However this is not the expected behaviour of the `AnimatedSwitcher` as described in its documentation:
> The same key can be used for a new child as was used for an already-outgoing child; the two will not be considered related. (For example, if a progress indicator with key A is first shown, then an image with key B, then another progress indicator with key A again, all in rapid succession, then the old progress indicator and the image will be fading out while a new progress indicator is fading in.)

It clearly states that multiple children with the same key can transition in and out at the same time.

My question now is: Am I allowed to remove this test, as it tests for _"wrong"_ behavior? Or how should I proceed?

jimgerth committed:
Revert "Fixed AnimatedSwitcher chain produced duplicates"

Revert the changes introduced by #107476, as they do not completely fix
the issue of duplicate keys in the AnimatedSwitcher switcher as
demonstrated by #121336, and introduce behavior that deviates from the
AnimatedSwitcher's documentation. Specifically, children that are
transitioning out are immediately removed from the widget tree, if
another child with the same key is added, while the AnimatedSwitcher is
clearly supposed to support multiple children with the same key
transitioning in and out at the same time according to its
documentation.

jimgerth committed:
Test for AnimatedSwitcher handling multiple children with the same key

Add a test for the AnimatedSwitcher, that test if multiple children with
the same key can be added and are properly transitioned in and out.

JimGerth commented:
I've written with @Hixie on discord and they told me to change (or replace in this case) the old test.
I eventually decided to entirely revert the previous PR #107476. See my previous comment, the PR description and the commit message for my reasoning.

goderbauer reviewed (approved):
LGTM, thanks for the fix!

christopherfujino commented:
> auto label is removed for flutter/flutter, pr: 121408, due to Validations Fail.

In cl/512895929 there's a slight goldens diff that could possibly be due to this change, WDYT @goderbauer?

goderbauer commented:
> In cl/512895929 there's a slight goldens diff that could possibly be due to this change

I triaged them as acceptable.
